src/ngraph/runtime/gpu/gpu_emitter.cpp

```diff
@@ -851,8 +851,7 @@ namespace ngraph
                         {
                             kernel::emit_memcpyDtD(writer, out[0], args[0]);
                         }
-                        // descriptors for tensors  with <= 4 dimensions
-                        else
+                        else if (out[0].get_shape().size() == 0)
```

Comment:
Is this intentionally left here to allow cudnn to handle the reduction to scalar case? I assume then it was faster than the cuda impl.? In general, do you have an understanding as to why cudnn is slower for the model you tested?

Reply:
Yes, It's intentionally left here, from my test, cudnn didn't do well for non-scalar case. And I will have another PR that optimized scalar case so we don't need cudnn for reduce anymore.

Reply:
It's still a bit surprising to me that its performance is much much worse. 4x improvement over cudnn reduce doesn't seem reasonable - even though I'm happy about it. Makes me wonder if there are some cudnnReduce options we missed originally.

Reply:
I did check the options, no something I'm aware.

Reply:
Case might be cudnn might be optimized for some case but not for others, and possible things may happen to our optimization. That's why we need an tuning framework.

Reply:
💯